cloudant-sync-datastore-core/src/main/java/com/cloudant/sync/internal/documentstore/DatabaseImpl.java

```diff
@@ -1191,4 +1159,43 @@ public DocumentRevision delete(final DocumentRevision rev) throws
         }
     }
 
+    @Override
+    public void createWithHistory(DocumentRevision revision, int revisionsStart, List<String>
+            revisionsIDs) throws DocumentException {
+
+        Map<String, Attachment> attachments = revision.getAttachments();
+        Collections.list(Collections.enumeration(revision.getAttachments().values()));
+
+        InternalDocumentRevision internalRev = new DocumentRevisionBuilder()
+                .setDocId(revision.getId())
+                .setRevId(revision.getRevision())
+                .setBody(revision.getBody())
+                .setAttachments(attachments)
+                .setDeleted(revision.isDeleted())
+                .build();
+
+        Map<String[], Map<String, PreparedAttachment>> preparedAttachments = Collections
+                .singletonMap
+                        (new String[]{revision.getId(), revision.getRevision()}, AttachmentManager
+                                .prepareAttachments(attachmentsDir, attachmentStreamFactory,
+                                        attachments));
+
+        // Couch _revisions.ids are newest -> oldest without a generation, so we need to manipulate
+        // them for forceInsert to an ascending order list with generational prefix
+        Map<String, Object> revisions = new HashMap<String, Object>(2);
+        revisions.put("start", revisionsStart);
+        revisions.put("ids", Collections.list(Collections.enumeration(revisionsIDs)));
+        // Parse into a descending order list of rev IDs
+        List<String> revIDs = CouchUtils.parseCouchDBRevisionHistory(Collections.<String, Object>
```

Comment:
Looks like you are the only caller of this method - before your changes it wasn't being used. Note also that when doing pull replication and force inserting, `DatastoreWrapper.bulkInsert` uses the similar `DocumentRevsUtils.createRevisionIdHistory`.

My suggestion is that we should try and share the logic between these code paths and avoid this call site having to put values into a map just to satisfy another method (for instance, it would be better to have some sort of `List<String> getHistory(String start, List<Integer> ids)`)